packages/perspective-viewer-d3fc/src/js/tooltip/generateHTML.js

```diff
@@ -0,0 +1,121 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2017, the Perspective Authors.
+ *
+ * This file is part of the Perspective library, distributed under the terms of
+ * the Apache License 2.0.  The full license can be found in the LICENSE file.
+ *
+ */
+import {select} from "d3";
+
+export function generateHtmlDefault(tooltipDiv, data, settings) {
+    console.log("data", data);
+    console.log("settings", settings);
+
+    // Add group data
+    if (settings.crossValues.length) {
+        const groups = data.crossValue.split ? data.crossValue.split("|") : [data.crossValue];
+        tooltipDiv
+            .select("#cross-values")
+            .selectAll("li")
+            .data(groups)
+            .join("li")
+            .each(eachValue(settings.crossValues));
+    }
+
+    const splitValues = getSplitValues(settings, data);
+    // Add data value
+    tooltipDiv
+        .select("#split-values")
+        .selectAll("li")
+        .data(splitValues.values)
+        .join("li")
+        .each(eachValue(splitValues.labels));
+
+    const dataValues = getDataValues(settings, data);
+
+    // Add data value
+    tooltipDiv
+        .select("#data-values")
+        .selectAll("li")
+        .data(dataValues.values)
+        .join("li")
+        .each(eachValue(dataValues.labels));
+}
+
+export function generateHtmlForHeatmap(tooltipDiv, data, settings) {
```

Comment:
As far as I can see, all that is really different between these two functions is the way we get the split and data values? Could we have alternate functions here rather than redefining the whole html generation function.

Reply:
The differences between the two functions is subtle, but it is important as it is tightly linked to the way the data for the chart is structured. The current single generateHTML function doesn't work for heatmap and will potentially get more complex as we add addition charts with different data structures. I have refactored the two functions so that they use a new addValuesFromData function that could be used in some situations. We could replace the standard tooltipDiv code with a function (see below), but I already think it is fairly concise and clear.
tooltipDiv
        .select("#split-values")
        .selectAll("li")
        .data(splitValues.values)
        .join("li")
        .each(eachValue(splitValues.labels));

Reply:
I would like to explore ways to create tooltip rendering code that works for both cases, but I'd be happy to merge this now and add something to the backlog to come back to it.

Reply:
I have raised IPJS-53 to address this.